### PR TITLE
More deterministic throttling

### PR DESCRIPTION
**Pull request by d3murphy**
Change to `ThrottledCallback` default behavior:
- Align publishes with a whole number multiple of the throttle period for more deterministic throttling.

Add optional behavior to `ThrottledMessageCallback`:
- Add option to use header timestamps
- Subclass will automatically select message header or time-based throttling based on passed flags
  - selection is both in `callback` and `operator()`

Integration
- Add config option to sensor models and set in callback on init

Add tests:
- Verify new default and header timestamp behavior
- Specific test to verify correct version of `operator()` is called

### CODE REVIEW

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 1]
```diff
@@ -148,32 +148,43 @@ class ThrottledCallback
...
-    // (c) The elpased time between now and the last called time is greater than the throttle period
+    // (a) The throttle period is zero, so we should always keep the callbacks
+    // (b) The elpased time between now and the last called time is greater than the throttle period
+    //  If this is the first time this has been called, just store what the most recent phase-aligned publish time
+    //  should have been so we can publish
     const ros::Time now = use_wall_time_ ? ros::Time(ros::WallTime::now().toSec()) : ros::Time::now();
-    if (last_called_time_.isZero() || throttle_period_.isZero() || now - last_called_time_ > throttle_period_)
+
+    bool keep = false;
+    if (last_called_time_.isZero() || throttle_period_.isZero() || now > last_called_time_ + throttle_period_)
```
Comment (d3murphy) [inline fuse_core/include/fuse_core/throttled_callback.h:168]: Condition on the right is the same, I just changed it because "a time that's after the next phase-aligned publish time" is more clear than "a time one period before the current time that is greater than the last phase-aligned publish time".

The previous condition was more clear for the old logic.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 2]
```diff
@@ -197,6 +208,7 @@ class ThrottledCallback
   bool use_wall_time_;             //<! The flag to indicate whether to use ros::WallTime or not
 
   ros::Time last_called_time_;  //!< The last time the keep callback was called
```
Comment (d3murphy) [inline fuse_core/include/fuse_core/throttled_callback.h:210]: Should maybe change name to `last_aligned_publish_time_`.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 3]
```diff
@@ -148,32 +148,43 @@ class ThrottledCallback
...
-      {
-        last_called_time_ += throttle_period_;
+      // Don't actually call the keep callback the first time, just figure out what the most recent phase-aligned
+      // output time would have been so we can compare future messages with it.
+      if (!last_called_time_.isZero()) {
+        keep = true;
       }
+
+      // Use the aligned publish time to avoid any accumulation of error
+      last_called_time_ = aligned_publish_time;
```
Comment (d3murphy) [inline fuse_core/include/fuse_core/throttled_callback.h]: Should maybe change the name of this member to `last_aligned_publish_time_`.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 4]
```diff
@@ -148,32 +148,43 @@ class ThrottledCallback
    *
    * @param[in] args The input arguments
    */
+
+  // TODO: Better determinism with this method
+  // [ ] 1) Use header timestamps
```
Comment (d3murphy) [inline fuse_core/include/fuse_core/throttled_callback.h:153]: Header timestamps would probably make a more significant improvement than the below.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 5]
```diff
@@ -148,32 +148,40 @@ class ThrottledCallback
...
+    // (b) The elpased time between now and the last called time is greater than the throttle period
+    //  If this is the first time this has been called, just store what the most recent phase-aligned publish time
+    //  should have been so we can publish next time.
     const ros::Time now = use_wall_time_ ? ros::Time(ros::WallTime::now().toSec()) : ros::Time::now();
-    if (last_called_time_.isZero() || throttle_period_.isZero() || now - last_called_time_ > throttle_period_)
+
+    bool keep = false;
+    if (throttle_period_.isZero()) {
+      keep = true;
+    } else if (last_called_time_.isZero() || now > last_called_time_ + throttle_period_)
```
Comment (d3murphy) [inline fuse_core/include/fuse_core/throttled_callback.h]: Condition on the right is the same, I just changed it because "a time that's after the next phase-aligned publish time" is more clear than "a time one period before the current time that is greater than the last phase-aligned publish time".

The previous condition was more clear for the old logic.

File: fuse_core/CMakeLists.txt [hunk 6]
```diff
@@ -146,7 +146,7 @@ if(CATKIN_ENABLE_TESTING)
   find_package(geometry_msgs REQUIRED)
 
   # Lint tests
-  set(ROSLINT_CPP_OPTS "--filter=-build/c++11,-runtime/references")
+  set(ROSLINT_CPP_OPTS "--filter=-build/c++20,-runtime/references")
```
Comment (efernandez) [inline fuse_core/CMakeLists.txt]: It looks like I forgot to updfate this. I think we should update that for all packages using some `sed` commands. If you want I can do that and send it on another MR

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 7]
```diff
@@ -190,7 +176,35 @@ class ThrottledCallback
...
+      const double time_s = time.toSec();
+      const double last_called_time_s =  time_s - std::fmod(time_s, throttle_period_.toSec());
+      last_called_time_.fromSec(last_called_time_s);
+    }
+
+    if (keep) {
+      if (keep_callback_) {
+        keep_callback_(std::forward<Args>(args)...);
+      }
+    } else if (drop_callback_)
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: We use this style instead - the code LGTM
```suggestion
    if (throttle_period_.isZero())
    {
      keep = true;
    }
    else if (last_called_time_.isZero() || time > last_called_time_ + throttle_period_)
    {
      if (!last_called_time_.isZero())
      {
        keep = true;
      }

      // Use the nearest phase-aligned multiple of the throttle period as our reference time to improve determinism
      const double time_s = time.toSec();
      const double last_called_time_s =  time_s - std::fmod(time_s, throttle_period_.toSec());
      last_called_time_.fromSec(last_called_time_s);
    }

    if (keep)
    {
      if (keep_callback_)
      {
        keep_callback_(std::forward<Args>(args)...);
      }
    }
    else if (drop_callback_)
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 8]
```diff
@@ -200,12 +214,37 @@ class ThrottledCallback
...
+
+  /**
+   * @brief Callback that uses message header timestamps to throttle the calls to the keep callback provided.
+   * When dropped because of throttling, the drop callback is called instead.
+   * @details The first message after the most recent whole-number multiple of the throttle_period_ is published
+   * and all others are dropped.
+   *
+   * @param[in] msg The input arguments
+   */
+  void message_stamp_callback(const typename M::ConstPtr& msg)
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: This is a good idea, but where is this method called?

I think we should also rename it to
```suggestion
  void messageStampCallback(const typename M::ConstPtr& msg)
```

And we should also check the message type `M` has `header.stamp` - you can use https://docs.ros.org/en/indigo/api/roscpp_traits/html/message__traits_8h_source.html#l00106 to `enable_if` this method or the entire subclass only for messages with header, which will have a stamp

**Review by efernandez — CHANGES_REQUESTED** [review]

File: fuse_core/CMakeLists.txt (hunk 6 above)
Comment (d3murphy) [reply fuse_core/CMakeLists.txt]: We should do in another MR. lmk if you want me to do it - either way is fine with me.

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 8 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h]: It's not called yet! I wanted to check the existing test cases with the new default logic and I was fighting to get the tests running. Now that they are running I will add a test to this module then look at adopting in the sensor models.

For adoption in the sensor models, we should probably add a config option to select which callback to use then update our configs to use the new one on our side after that goes in. I'm guessing we will also have to update some expected test results. Does that sound like a reasonable approach?

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 7 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h]: I keep forgetting this :-). Having a hard time without my auto-formatter.

File: fuse_core/CMakeLists.txt [hunk 9]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.16.3)
```
Comment (d3murphy) [inline fuse_core/CMakeLists.txt]: Should we update all modules with this as well? This is the one we use internally.

File: fuse_core/CMakeLists.txt (hunk 6 above)
Comment (efernandez) [reply fuse_core/CMakeLists.txt]: IMO it should be on another MR

File: fuse_core/CMakeLists.txt (hunk 9 above)
Comment (efernandez) [reply fuse_core/CMakeLists.txt]: If that's required for some reason, yes. Let's update all to that version - it should ideally be a change in another MR, at least the one changing all packages

Comment (d3murphy) [reply fuse_core/CMakeLists.txt]: Will do - leaving here for now so I can iterate on tests locally. Will remove before I mark ready for review.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 10]
```diff
@@ -145,32 +146,69 @@ class ThrottledCallback
...
+
+protected:
+  /**
+   * @brief The implementation of the throttled callback
+   *
+   * @param[in] time The time used for throttling
+   * @param[in] args The input arguments
+   */
+  template <class... Args>
+  void callbackImpl(const ros::Time time, Args&&... args)
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: Any reason we don't take `ros::Time` by reference?
```suggestion
  void callbackImpl(const ros::Time& time, Args&&... args)
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 11]
```diff
@@ -57,14 +57,15 @@ class ThrottledCallback
...
    * @brief Constructor
    *
-   * @param[in] keep_callback   The callback to call when kept, i.e. not dropped. Defaults to nullptr
-   * @param[in] drop_callback   The callback to call when dropped because of the throttling. Defaults to nullptr
-   * @param[in] throttle_period The throttling period duration in seconds. Defaults to 0.0, i.e. no throttling
-   * @param[in] use_wall_time   Whether to use ros::WallTime or not. Defaults to false
+   * @param[in] keep_callback       The callback to call when kept, i.e. not dropped. Defaults to nullptr
+   * @param[in] drop_callback       The callback to call when dropped because of the throttling. Defaults to nullptr
+   * @param[in] throttle_period     The throttling period duration in seconds. Defaults to 0.0, i.e. no throttling
+   * @param[in] use_wall_time       Whether to use ros::WallTime or not. Defaults to false
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: Why the extra indentation?
```suggestion
   * @param[in] keep_callback   The callback to call when kept, i.e. not dropped. Defaults to nullptr
   * @param[in] drop_callback   The callback to call when dropped because of the throttling. Defaults to nullptr
   * @param[in] throttle_period The throttling period duration in seconds. Defaults to 0.0, i.e. no throttling
   * @param[in] use_wall_time   Whether to use ros::WallTime or not. Defaults to false
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 12]
```diff
@@ -57,14 +57,15 @@ class ThrottledCallback
...
    */
   ThrottledCallback(Callback&& keep_callback = nullptr,  // NOLINT(whitespace/operators)
                     Callback&& drop_callback = nullptr,  // NOLINT(whitespace/operators)
-                    const ros::Duration& throttle_period = ros::Duration(0.0), const bool use_wall_time = false)
+                    const ros::Duration& throttle_period = ros::Duration(0.0),
+                    const bool use_wall_time = false)
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: Why split lines?
```suggestion
                    const ros::Duration& throttle_period = ros::Duration(0.0), const bool use_wall_time = false)
```

I don't really have a strong opinion on this one, but this isn't really required for this change

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 13]
```diff
@@ -145,32 +146,69 @@ class ThrottledCallback
...
 
-      if (last_called_time_.isZero())
-      {
-        last_called_time_ = now;
-      }
-      else
+      // Use the nearest phase-aligned multiple of the throttle period as our reference time to improve determinism
+      const double time_s = time.toSec();
+      const double last_called_time_s =  time_s - std::fmod(time_s, throttle_period_.toSec());
+      last_called_time_.fromSec(last_called_time_s);
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h:203]: If `time < throttle_period_` then `std::fmod(time_s, throttle_period_.toSec())` would be the same as `time_s` and `last_called_time_` would be `0`. Maybe we could handle that as a special case, but this should actually work. That's fine.

With that in mind, eventually we'll get a `time` that is `> throttle_period_`. It looks like in that case we'll still skip the message. We'll set `last_called_time_` to `throttle_period_`.

The next messages we'll get will also be filtered out until we get a `time` that's `> 2 * throttle_period_`.

If think we should also keep the message in that case. I believe we should use flag to detect the first call instead of relying on `last_called_time_.isZero()`.

Maybe something like this (github doesn't support suggestions on deleted lines):
```c++
else if (!last_call_time_set_ || time > last_called_time_ + throttle_period_)
{
      keep = last_call_time_set_;

      // Use the nearest phase-aligned multiple of the throttle period as our reference time to improve determinism
      const double time_s = time.toSec();
      const double last_called_time_s =  time_s - std::fmod(time_s, throttle_period_.toSec());
      last_called_time_.fromSec(last_called_time_s);
      last_call_time_set_ = true;
    }
```
where `last_call_time_set_` is initialized to `false`

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 14]
```diff
@@ -179,8 +217,60 @@ class ThrottledCallback
     }
   }
 
+  Callback keep_callback_;         //!< The callback to call when kept, i.e. not dropped
+  Callback drop_callback_;         //!< The callback to call when dropped because of throttling
+  ros::Duration throttle_period_;  //!< The throttling period duration in seconds
+  bool use_wall_time_;             //<! The flag to indicate whether to use ros::WallTime or not
+
+  ros::Time last_called_time_;  //!< The last time the keep callback was called
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]:
```suggestion
  bool last_called_time_set_{ false };  //!< The flag to indicate whether the last time the keep callback was called has been set or not yet
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 15]
```diff
@@ -179,8 +217,60 @@ class ThrottledCallback
...
+
+  ThrottledMessageCallback(Callback&& keep_callback = nullptr,  // NOLINT(whitespace/operators)
+                           Callback&& drop_callback = nullptr,  // NOLINT(whitespace/operators)
+                           const ros::Duration& throttle_period = ros::Duration(0.0),
+                           const bool use_wall_time = false,
+                           const bool use_header_timestamps = false)
+    : ThrottledCallback<Callback>(std::move(keep_callback), std::move(drop_callback), throttle_period,
+                                  use_wall_time), use_header_timestamps_(use_header_timestamps)
+    {
+      if (use_wall_time && use_header_timestamps) {
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: Style
```suggestion
      if (use_wall_time && use_header_timestamps)
      {
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 16]
```diff
@@ -179,8 +217,60 @@ class ThrottledCallback
...
+        throw std::runtime_error("Cannot specify both wall time and header timestamps for callback throttling. Use only one.");
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: I wonder if it'd make more sense to turn these flags into template parameters :thinking: 

I'm just wondering. It could a separate MR if you think it'd make things cleaner.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 17]
```diff
@@ -179,8 +217,60 @@ class ThrottledCallback
...
+  {
+    if (!use_header_timestamps_)
+    {
+      // Use base-class version
+      ThrottledCallback<Callback>::callback(std::forward<Args>(args)...);
+    }
+    else
+    {
+      messageStampCallback(std::forward<Args>(args)...);
+    }
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: I'd prefer this
```suggestion
    if (use_header_timestamps_)
    {
      messageStampCallback(std::forward<Args>(args)...);
    }
    else
    {
      // Use base-class version
      ThrottledCallback<Callback>::callback(std::forward<Args>(args)...);
    }
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 18]
```diff
@@ -190,22 +280,40 @@ class ThrottledCallback
...
+   * When dropped because of throttling, the drop callback is called instead.
+   * @details The first message after the most recent whole-number multiple of the throttle_period_ is published
+   * and all others are dropped.
+   *
+   * @param[in] msg The message to be throttled
+   */
+  template <typename = typename std::enable_if_t<ros::message_traits::HasHeader<M>::value>>
+  void messageStampCallback(const typename M::ConstPtr& msg)
+  {
+    const auto msg_stamp_time = ros::Time(msg->header.stamp.sec, msg->header.stamp.nsec);
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]:
```suggestion
    const ros::Time msg_stamp_time(msg->header.stamp.sec, msg->header.stamp.nsec);
```
or
```suggestion
    const ros::Time msg_stamp_time{ msg->header.stamp.sec, msg->header.stamp.nsec };
```
depending on what's more common in `fuse`, which I believe it's the second case, but not sure.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 19]
```diff
@@ -190,22 +280,40 @@ class ThrottledCallback
...
+  }
+
+  /**
+   * @brief Use header timestamps flag setter
+   *
+   * @param[in] use_wall_time Whether to use msg header timestamps or not
+   */
+  void setUseHeaderTimestamps(const bool use_header_timestamps)
+  {
+    if (use_header_timestamps && this->use_wall_time_) {
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]:
```suggestion
    if (use_header_timestamps && this->use_wall_time_)
    {
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 20]
```diff
@@ -78,41 +78,60 @@ class PointPublisher
...
+      geometry_msgs::PointStamped point_message;
+      point_message.point.x = i;
 
       publisher_.publish(point_message);
 
       rate.sleep();
     }
   }
 
+  void publishWithHeaderTimestamp(const ros::Time timestamp)
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]:
```suggestion
  void publishWithHeaderTimestamp(const ros::Time& timestamp)
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 21]
```diff
@@ -78,41 +78,60 @@ class PointPublisher
...
+  {
+    ros::WallTime subscriber_timeout = ros::WallTime::now() + ros::WallDuration(1.0);
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]:
```suggestion
    const ros::WallTime subscriber_timeout = ros::WallTime::now() + ros::WallDuration(1.0);
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 22]
```diff
@@ -78,41 +78,60 @@ class PointPublisher
...
+
+ static_assert(ros::message_traits::HasHeader<geometry_msgs::PointStamped>::value == true);
 class PointSensorModel
 {
 public:
   /**
    * @brief Constructor
    *
-   * @param[in] throttle_period The throttle period duration in seconds
+   * @param[in] throttle_period  The throttle period duration in seconds
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]:
```suggestion
   * @param[in] throttle_period The throttle period duration in seconds
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 23]
```diff
@@ -78,41 +78,60 @@ class PointPublisher
...
 /**
- * @brief A dummy point sensor model that uses a fuse_core::ThrottledMessageCallback<geometry_msgs::Point> with a keep
+ * @brief A dummy point sensor model that uses a fuse_core::ThrottledMessageCallback<geometry_msgs::PointStamped> with a keep
  * and drop callback.
  *
  * The callbacks simply count the number of times they are called, for testing purposes. The keep callback also caches
  * the last message received, also for testing purposes.
  */
+
+ static_assert(ros::message_traits::HasHeader<geometry_msgs::PointStamped>::value == true);
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]:
```suggestion
static_assert(ros::message_traits::HasHeader<geometry_msgs::PointStamped>::value == true);
```

I didn't know you could have this here. I thought it had to be in a function or method scope.

File: fuse_core/test/test_throttled_callback.cpp [hunk 24]
```diff
@@ -227,35 +257,128 @@ TEST(ThrottledCallback, DropMessagesIfThrottlePeriodIsGreaterThanPublishPeriod)
...
+  point_msg.header.stamp += ros::Duration().fromSec(1.0);
+  msg_const = boost::make_shared<const geometry_msgs::PointStamped>(point_msg);
+  message_header_calback(msg_const);
+
+  // The header callback should keep this one because the header timestamp is advanced more than the throttle period
+  EXPECT_EQ(header_timestamp_kept, 1);
+}
+
+
+
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]:
```suggestion


```

File: fuse_core/test/test_throttled_callback.cpp [hunk 25]
```diff
@@ -227,35 +257,128 @@ TEST(ThrottledCallback, DropMessagesIfThrottlePeriodIsGreaterThanPublishPeriod)
...
 int main(int argc, char** argv)
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp:377]: I think we should also test the case to `time` being `< throttle_period_`, which it's probably not covered by the unit tests.

**Review by efernandez — CHANGES_REQUESTED** [review]

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 13 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h:203]: I was thinking this isn't really a valid use-case. Even if the system comes up at unix epoch, it will take time for everything to start and it doesn't seem realistic to me that we care about data for the first fraction of a second after the epoch. If it is a valid use-case, I can implement the requested changes.

File: fuse_core/test/test_throttled_callback.cpp (hunk 21 above)
Comment (d3murphy) [reply fuse_core/test/test_throttled_callback.cpp]: Copied from your implementation above :smile:

File: fuse_core/test/test_throttled_callback.cpp (hunk 23 above)
Comment (d3murphy) [reply fuse_core/test/test_throttled_callback.cpp]: This is how I usually use them. I kind of prefer this to `std::enable_if` which is cool but smells a little bit too much like magic to me.

File: fuse_core/test/test_throttled_callback.cpp (hunk 25 above)
Comment (d3murphy) [reply fuse_core/test/test_throttled_callback.cpp:382]: It's definitely not. Again not sure if that's a valid use-case

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 16 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h]: I don't think we can because these are read from param file at runtime.

File: fuse_core/test/test_throttled_callback.cpp (hunk 25 above)
Comment (d3murphy) [reply fuse_core/test/test_throttled_callback.cpp:377]: Per offline discussion: Will do

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 13 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h:203]: Per offline discussion, adopted recommended changes since it's not a huge increase in complexity and could be valid in sim.

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 12 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h]: Yeah fair. I probably modified this signature at some point and reverted it back without going back to the old formatting. As for style, I think I just liked the consistency - Pretty sure that clang-format does this where once it starts to split them across lines, each one gets its own line. Can go back to the old style if preferred.

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 13 above)
Comment (efernandez) [reply fuse_core/include/fuse_core/throttled_callback.h:203]: It's a corner case that could happen in simulation, where time starts from zero. Likely a corner case because it could take time to start things up, but if the filter starts running right away and simulation is running slower than real time, it could be possible

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 12 above)
Comment (efernandez) [reply fuse_core/include/fuse_core/throttled_callback.h]: I don't have a strong opinion on the style here, just trying to reduce the changes to the scope of the PR motivation

Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h]: Ok understood, I'll revert this to keep the file history clean.

File: fuse_core/include/fuse_core/throttled_callback.h (hunk 10 above)
Comment (d3murphy) [reply fuse_core/include/fuse_core/throttled_callback.h]: This is a micro (more like nano!) optimization since sizeof(time) == 8, so it's probably infinitesimally faster to pass by value. Can change to reference since I see that's the convention.

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 26]
```diff
@@ -179,8 +217,62 @@ class ThrottledCallback
...
+                           const ros::Duration& throttle_period = ros::Duration(0.0),
+                           const bool use_wall_time = false,
+                           const bool use_header_timestamps = false)
+    : ThrottledCallback<Callback>(std::move(keep_callback), std::move(drop_callback), throttle_period,
+                                  use_wall_time), use_header_timestamps_(use_header_timestamps)
+    {
+      if (use_wall_time && use_header_timestamps)
+      {
+        throw std::runtime_error("Cannot specify both wall time and header timestamps for callback throttling."
+                                 "Use only one.");
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: You're missing a blank space after the dot `.`
```suggestion
        throw std::runtime_error("Cannot specify both wall time and header timestamps for callback throttling."
                                 " Use only one.");
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 27]
```diff
@@ -179,8 +217,62 @@ class ThrottledCallback
...
+      }
+    }
+
+ /**
+ * @brief Callback that selects the correct timestamp to throttle the calls to the keep callback provided based
+ * the configuration arguments.
+ * @param[in] args The callback args, will be forwarded to the callback
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: Add empty line for consistency with other doxygen blocks
```suggestion
 * the configuration arguments.
 *
 * @param[in] args The callback args, will be forwarded to the callback
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 28]
```diff
@@ -190,22 +282,42 @@ class ThrottledCallback
...
+  void messageStampCallback(const typename M::ConstPtr& msg)
+  {
+    const ros::Time msg_stamp_time(msg->header.stamp.sec, msg->header.stamp.nsec);
+    this->callbackImpl(msg_stamp_time, msg);
+  }
+
+  /**
+   * @brief Use header timestamps flag setter
+   *
+   * @param[in] use_wall_time Whether to use msg header timestamps or not
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]: The param name is wrong
```suggestion
   * @param[in] use_header_timestamps Whether to use msg header timestamps or not
```

File: fuse_core/include/fuse_core/throttled_callback.h [hunk 29]
```diff
@@ -190,22 +282,42 @@ class ThrottledCallback
...
+   */
+  void setUseHeaderTimestamps(const bool use_header_timestamps)
+  {
+    if (use_header_timestamps && this->use_wall_time_)
+    {
+      throw std::runtime_error("Cannot specify both wall time and header timestamps for callback throttling."
+                               "Use only one.");
```
Comment (efernandez) [inline fuse_core/include/fuse_core/throttled_callback.h]:
```suggestion
      throw std::runtime_error("Cannot specify both wall time and header timestamps for callback throttling."
                               " Use only one.");
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 30]
```diff
@@ -141,18 +158,29 @@ class PointSensorModel
...
+  const geometry_msgs::PointStamped::ConstPtr getLastKeptMessage() const
   {
     return last_kept_message_;
   }
 
+    /**
+   * @brief The last message kept getter
+   *
+   * @return The last message kept. It would be nullptr if no message has been kept so far
+   */
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]: Wrong doxygen
```suggestion
  /**
   * @brief Reset the number of kept and dropped messages
   */
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 31]
```diff
@@ -227,35 +255,127 @@ TEST(ThrottledCallback, DropMessagesIfThrottlePeriodIsGreaterThanPublishPeriod)
...
+
+  for (; i < 2 * kPublishToThrottleRatio; ++i)
+  {
+    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
+    publisher.publishWithHeaderTimestamp(timestamp);
+  }
+  ros::Duration(0.1).sleep();
+
+  EXPECT_EQ(sensor_model.getKeptMessages(), 1);
+  EXPECT_EQ(sensor_model.getDroppedMessages(), kPublishToThrottleRatio - 1);
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]: I think it'd be cleaner if you don't use `i` in the `EXPECT_EQ` - you can use `kPublishToThrottleRatio` instead
```suggestion
  sensor_model.reset();

  timestamp.fromSec(kBaseTimeS - kEpsilonS);

  for (std::size_t i = 0; i < kPublishToThrottleRatio; ++i)
  {
    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
    publisher.publishWithHeaderTimestamp(timestamp);
  }

  ros::Duration(0.1).sleep();
  EXPECT_EQ(sensor_model.getKeptMessages(), 0);
  EXPECT_EQ(sensor_model.getDroppedMessages(), kPublishToThrottleRatio);

  // Publish for another period, phase aligned to just before the period division
  // Should keep one message
  sensor_model.reset();

  for (std::size_t i = 0; i < kPublishToThrottleRatio; ++i)
  {
    timestamp.fromSec(timestamp.toSec() + kPublishPeriodS);
    publisher.publishWithHeaderTimestamp(timestamp);
  }
  ros::Duration(0.1).sleep();

  EXPECT_EQ(sensor_model.getKeptMessages(), 1);
  EXPECT_EQ(sensor_model.getDroppedMessages(), kPublishToThrottleRatio - 1);
```

File: fuse_core/test/test_throttled_callback.cpp [hunk 32]
```diff
@@ -227,35 +255,127 @@ TEST(ThrottledCallback, DropMessagesIfThrottlePeriodIsGreaterThanPublishPeriod)
...
+  // Call with the same message again
+  base_callback(msg_const);
+  message_header_calback(msg_const);
+  // The base callback should keep this one because it's related to elapsed time
+  EXPECT_EQ(base_kept, 1);
+  // The header callback should not keep this one because the header timestamp is the same as the old one
+  EXPECT_EQ(header_timestamp_kept, 0);
+
+  // Advance the header timestamp and check that the header timestamp based throttle keeps it
+  point_msg.header.stamp += ros::Duration().fromSec(1.0);
```
Comment (efernandez) [inline fuse_core/test/test_throttled_callback.cpp]: No need to call `fromSec`
```suggestion
  point_msg.header.stamp += ros::Duration(1.0);
```

**Review by efernandez — CHANGES_REQUESTED** [review]
LGTM - but I have some minor comments and change requests

I also wonder if the unit tests evaluate the case of the time starting from zero - I think they do, right?

**Review by efernandez — APPROVED** [review]
